src/parser/mod.rs

```diff
@@ -471,9 +471,47 @@ fn term<'a, 'b>(input: Input<'a, 'b>) -> AstResult<'a, 'b> {
 parser!(function_call loc =
     function <- member_access;
     args <- many1(function_argument);
-    Ast::function_call(function, args, loc)
+    desugar_explicit_currying(function, args, loc)
 );
 
+/// Turns (foo _  _ 2) into (\ _$1 _$2 . (foo _$1 _$2 2))
+fn desugar_explicit_currying<'a>(function: Ast<'a>, args: Vec<Ast<'a>>, loc: Location<'a>) -> Ast<'a> {
+    if matches_not_typeconstructor(&function) && args.iter().any(matches_underscore) {
+        return curried_function_call(function, args, loc)
+    }
+
+    Ast::function_call(function, args, loc)
+}
+
+fn curried_function_call<'a>(function: Ast<'a>, args: Vec<Ast<'a>>, loc: Location<'a>) -> Ast<'a> {
+    let mut curried_args = vec![];
+    let mut curried_arg_count = 0;
+    let args: Vec<Ast<'a>> = args
+        .into_iter()
+        .map(|arg| {
+            if matches_underscore(&arg) {
+                curried_arg_count += 1;
+                let curried_arg = format!("_${}", curried_arg_count);
```

Comment:
Nitpick: Can we get rid of the `_` in front of the generated names? That way it'd match the `$1, $2, ..., $n` syntax used by some other languages with their lambda syntaxes.